### PR TITLE
Suggestion of refactor around comparison to nan & infinity.

### PR DESCRIPTION
**Pull request by smoothdeveloper**
Notes:
* FSharp.Stats.Testing.SAM, the current implementation only checks for positive infinity, just wondering if this is an overlook (sorry, not familiar with the underlying domain concepts), might be worth a comment if it is not
* such a refactor will help with https://github.com/fable-compiler/Fable/issues/3489

* [ ] The project builds without problems on your machine
* [ ] Added unit tests regarding the added features

For first point, I had to tweak `INormFloat<bignum>` interface implementation in Numeric.fs, as it didn't compile in my environment (I think the implementation is plain wrong).

For the second point, this is a refactoring not supposed to add any APIs nor to change any results, all the tests are passing.

Related: https://github.com/fable-compiler/Fable/issues/3489

### CODE REVIEW

File: src/FSharp.Stats/Testing/SAM.fs
Comment (bvenn) [inline src/FSharp.Stats/Testing/SAM.fs:1]: @zieglerSe could you check, if negative infinity is a case that should be covered?


```fsharp
//example that result in -infinity
(-10. / 0.) // gives -infinity
(10. / -0.) // gives -infinity
```

Comment (bvenn) [reply src/FSharp.Stats/Testing/SAM.fs:1]: As discussed in #273 it can be changed to `isInf` 👍

**Review by bvenn — CHANGES_REQUESTED** [review]
change isPosInf -> isInf as discussed in #273

File: src/FSharp.Stats/Testing/SAM.fs
Comment (smoothdeveloper) [reply src/FSharp.Stats/Testing/SAM.fs:1]: @bvenn @zieglerSe I tried using the SAM module, it doesn't seem to behave differently by adding the extra check, if a sample contains either `-infinity`, `infinity` or `nan`, the results won't be usable as the fields that contextualizes the sample among the others will be `nan`, and the `*SigBioitem` will be empty.

image

I think it would be good to ponder on the semantics that modules want to have, in terms of handling those non finite values, but I'm not a domain expert.

In "ideal world", I'd want algorithms that compose the statistical functions, to have a "statically verifiable" behaviour in terms of what those non finite values do.

* if all is to be baked statistically, having attributes (say `PropagatesNonFiniteValues` or `DiscardsNonFiniteValues` or `FailsOnNonFiniteValues`), and unit tests that would verify if these hold true
* or if we want this to be a runtime choice (which I think is nice to consider), the algorithms would have a default behaviour advertised in their documentation, and honour a runtime preference, by calling a preprocessing function over the data, or some other solution to this design problem

I don't have the perspective of what end user wants, and also, in terms of making it tractable for contributors, to consider those aspects when new modules are added.

cc: @charlesroddie

Comment (smoothdeveloper) [reply src/FSharp.Stats/Testing/SAM.fs:1]: @bvenn I've updated SAM to use `isInf`, thanks for the feedback!

Comment (bvenn) [reply src/FSharp.Stats/Testing/SAM.fs:1]: For context: The check for non finite values that this PR tackles is not referring to the initial values. Internally, a statistic is calculated by using both input collections. Even if these contain valid/finite data, the statistic can become non finite. This is absolutely okay, but must be taken care of in subsequent calculations, which is done by this exact filter step.


I'm not quite sure if I get your point completely, so let me rephrase your suggestions:

If the input data for SAM contains non finite values, currently the result is empty as you showed. This is valid because SAM cannot report significant bioitems if one contains abundances that are non finite. You suggest an optional parameter that defines how to proceed if the input contains non-finite values. In theory this would be a union type that defines different paths to follow for many of the analysis methods within FSharp.Stats:

```fsharp
type HandleNonFiniteValues =
    | PropagatesNonFiniteValues
    | DiscardsNonFiniteValues 
    | FailsOnNonFiniteValues
    | ...


module SAM
    let twoClassUnpaired iterations fdr data1 data2 rnd (handleNaN: HandleNonFiniteValues) =
        let data1Filtered,data2Filtered = 
            match handleNaN with
                | FailsOnNonFiniteValues -> //if data1 or data2 contains nan|infinity then failwithf
                | PropagatesNonFiniteValues -> data1,data2
                | DiscardsNonFiniteValues ->  
                    Seq.zip data1 data2
                    |> Seq.filter (fun (a,b) -> //filter if a and b contains finite data
                    |> Seq.unzip
                | _ -> ...
        // proceed with SAM

```

While I like the gained convenience, I see two problems with this procedure:

1. There are various ways to deal with these cases throughout the library. There are a ton of ways how to deal with this situation and each must be carefully tackled. The data could be filtered, imputed by KNN or random, replaced by 0., or somehow transformed. I'm not entirely sure if the various options to choose from would make the life easier for the user. 
A default value would be a way to deal with it, but I don't think there is a sensible default procedure for most of our methods (including SAM).
2. The bigger problem I see is the inability for the user to see problems with the initial data. In data science I try to always have the control over my data and frequently check the current status. You would want to know if there is any data that don't meet the expectations for a methodology. 
If the SAM-result is as you showed, it is clear, that there is something going on with the input data. As analyst you now should check your data and do the filtering yourself before starting new SAM attempt. I personally think of it as advantage to be in full control of the data and don't have the data filtered internally using defaults.

However, I think it is a great idea to provide multiple ways to deal with `nan`s in the inputs! But I think these should be specialised for each workflow with a sensible collection of applicable filters/transformations. Additionally, the parameter should either be mandatory so the user is in full control, or the default should be to **not**  transform or filter the inputs in any way.

@smoothdeveloper is this what you meant?

**Review by bvenn — APPROVED** [review]

File: src/FSharp.Stats/Testing/SAM.fs
Comment (smoothdeveloper) [reply src/FSharp.Stats/Testing/SAM.fs:1]: @bvenn thanks for digging into those considerations, and clarifying that algorithms can definitely come up with non finite values, even with inputs that are "sanitized".

Overall, the point I was bringing is more a general design concern, not specific to SAM, so I am not suggesting to move to change anything, but yes, your code example 100% grasps the idea that there could be different behaviours that can be useful.

I'll let this sit for a bit and consider how to bring it as a discussion topic, but you'd probably do a better job at initiating such discussion, especially if you bring actual examples where you'd like to use the same function in different contexts, to handle those non finite values in data exploration, or production.

I also feel, to address such concern the best, we'd need some F# wizards, I could see for example that computation expression could be leveraged to specify the intent and give the library an "ambient context".

From language perspective, it makes me think of C# handling of overflow with the `checked` and `unchecked` keywords https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/checked-and-unchecked
